### PR TITLE
neofs: allow to register multiple candidates per block

### PR DESCRIPTION
**Pull request by fyrchik**
Related #101.

### CODE REVIEW

File: neofs/neofs_contract.go [hunk 1]
```diff
@@ -212,24 +220,26 @@ func InnerRingCandidateAdd(key interop.PublicKey) {
 	}
 
 	c := common.IRNode{PublicKey: key}
-	candidates := getNodes(ctx, candidatesKey)
-
-	list, ok := addNode(candidates, c)
```
Comment (alexvanin) [inline neofs/neofs_contract.go:217]: `addNode` function can be removed.

File: neofs/neofs_contract.go [hunk 2]
```diff
@@ -568,6 +578,17 @@ func setConfig(ctx storage.Context, key, val interface{}) {
 	storage.Put(ctx, storageKey, val)
 }
 
+func hasNode(ctx storage.Context, key []byte, n common.IRNode) bool {
+	rawNode := std.Serialize(n)
+	it := storage.Find(ctx, key, storage.ValuesOnly)
+	for iterator.Next(it) {
+		if common.BytesEqual(rawNode, iterator.Value(it).([]byte)) {
+			return true
```
Comment (alexvanin) [inline neofs/neofs_contract.go:586]: Why `hasNode` compares serialized `common.IRNode{}` structures? It seems that now contract stores only public keys of those nodes. I suppose it should be 

```go
prefix := []byte(candidatesKey)
stKey := append(prefix, n.PublicKey...)
return storage.Get(ctx, stKey) != nil
```

File: neofs/neofs_contract.go [hunk 3]
```diff
@@ -212,24 +220,26 @@ func InnerRingCandidateAdd(key interop.PublicKey) {
...
 
 	transferred := gas.Transfer(from, to, fee, []byte(ignoreDepositNotification))
 	if !transferred {
 		panic("irCandidateAdd: failed to transfer funds, aborting")
 	}
 
+	stKey := append([]byte(candidatesKey), key...)
+	if storage.Get(ctx, stKey) != nil {
+		panic("irCandidateAdd: candidate already in the list")
+	}
```
Comment (alexvanin) [inline neofs/neofs_contract.go:239]: Considering comment about `hasNode`, I am a bit confused. Do we need to check it there if `hasNode` should do it instead? Maybe we do not need `hasNode` at all with this check?

File: neofs/neofs_contract.go (hunk 2 above)
Comment (fyrchik) [reply neofs/neofs_contract.go:586]: Fixed. My initial implementation was governed by the fact that `InnerRingCandidateAdd` can, in theory, update node info if the `IRNode` struct will be extended in future. Though old one compared public keys only so we are ok I thing.

File: neofs/neofs_contract.go (hunk 3 above)
Comment (fyrchik) [reply neofs/neofs_contract.go]: Removed.

File: neofs/neofs_contract.go [hunk 4]
```diff
@@ -172,16 +187,11 @@ func InnerRingCandidateRemove(key interop.PublicKey) {
...
-		if !common.BytesEqual(c.PublicKey, key) {
-			nodes = append(nodes, c)
-		} else {
-			runtime.Log("irCandidateRemove: candidate has been removed")
-		}
+	prefix := []byte(candidatesKey)
+	stKey := append(prefix, key...)
+	if storage.Get(ctx, stKey) != nil {
+		storage.Delete(ctx, stKey)
+		runtime.Log("irCandidateRemove: candidate has been removed")
```
Comment (alexvanin) [inline neofs/neofs_contract.go:194]: This code should be executed after `common.RemoveVotes(ctx, hashID)`

For notary disabled environment, there is a pattern to prepare data before (to decrease extra GAS addition) `common.RemoveVotes(ctx, hashID)` but store in in the contract memory after. This is way we were preparing new list there, but actually update it (`common.SetSerialized()`) in the end.

File: neofs/neofs_contract.go (hunk 1 above)
Comment (alexvanin) [reply neofs/neofs_contract.go:218]: Still there!

File: neofs/neofs_contract.go (hunk 2 above)
Comment (alexvanin) [reply neofs/neofs_contract.go:582]: `hasNode` has no references now, remove?

**Review by alexvanin — CHANGES_REQUESTED** [review]

File: neofs/neofs_contract.go (hunk 4 above)
Comment (fyrchik) [reply neofs/neofs_contract.go:194]: Fixed.

**Review by alexvanin — APPROVED** [review]

**Comment by alexvanin** [comment]
Rebase, please.

**Pull request merged**